[JIT] Remove profile nodes before BatchMM.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44073 [TensorExpr] Fuser: do not fuse ops with 0-dim tensors.
* **#43961 [JIT] Remove profile nodes before BatchMM.**

Currently we're removing prim::profile nodes and embed the type info
directly in the IR right before the fuser, because it is difficult to
fuse in a presence of prim::profile nodes. It turns out that BatchMM has
a similar problem: it doesn't work when there are prim::profile nodes in
the graph. These two passes run next to each other, so we could simply
remove prim::profile nodes slightly earlier: before the BatchMM pass.

Differential Revision: [D23453266](https://our.internmc.facebook.com/intern/diff/D23453266)

Commit:
[JIT] Remove profile nodes before BatchMM.

Currently we're removing prim::profile nodes and embed the type info
directly in the IR right before the fuser, because it is difficult to
fuse in a presence of prim::profile nodes. It turns out that BatchMM has
a similar problem: it doesn't work when there are prim::profile nodes in
the graph. These two passes run next to each other, so we could simply
remove prim::profile nodes slightly earlier: before the BatchMM pass.

[ghstack-poisoned]

Commit:
rebase on "[JIT] Remove profile nodes before BatchMM."


Currently we're removing prim::profile nodes and embed the type info
directly in the IR right before the fuser, because it is difficult to
fuse in a presence of prim::profile nodes. It turns out that BatchMM has
a similar problem: it doesn't work when there are prim::profile nodes in
the graph. These two passes run next to each other, so we could simply
remove prim::profile nodes slightly earlier: before the BatchMM pass.

Differential Revision: [D23453266](https://our.internmc.facebook.com/intern/diff/D23453266)

[ghstack-poisoned]

eellison commented on torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:
```diff
@@ -224,7 +224,12 @@ void runDiffGraphPasses(std::shared_ptr<Graph>& graph) {
     // and must be removed for fusion.
     LowerSimpleTuples(graph);
-    GRAPH_DUMP("After LowerSimpleTuples, before BatchMM", graph);
+    GRAPH_DUMP(
```
Can you add a comment abt why we're calling `RemoveProfileNodesAndSpecializeTypes`, and that any pass that uses specialized tensor types has to be run after  `FuseTensorExprs`.

eellison commented on torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:
```diff
@@ -224,7 +224,12 @@ void runDiffGraphPasses(std::shared_ptr<Graph>& graph) {
+    GRAPH_DUMP(
+        "After LowerSimpleTuples, before RemoveProfileNodesAndSpecializeTypes",
+        graph);
+    RemoveProfileNodesAndSpecializeTypes(graph);
```
Can we call `EraseShapeInformation` if `!tensorExprFuserEnabled()`

ZolotukhinM replied:
I moved here a call to `RemoveTensorTypeSpecializations` from the TE fuser - that should be equivalent, right?

eellison reviewed (approved):
LGTM

Commit:
Update on "[JIT] Remove profile nodes before BatchMM."


Currently we're removing prim::profile nodes and embed the type info
directly in the IR right before the fuser, because it is difficult to
fuse in a presence of prim::profile nodes. It turns out that BatchMM has
a similar problem: it doesn't work when there are prim::profile nodes in
the graph. These two passes run next to each other, so we could simply
remove prim::profile nodes slightly earlier: before the BatchMM pass.

Differential Revision: [D23453266](https://our.internmc.facebook.com/intern/diff/D23453266)

[ghstack-poisoned]

Krovatkin reviewed (approved):
:shipit:

Commit:
Update on "[JIT] Remove profile nodes before BatchMM."


Currently we're removing prim::profile nodes and embed the type info
directly in the IR right before the fuser, because it is difficult to
fuse in a presence of prim::profile nodes. It turns out that BatchMM has
a similar problem: it doesn't work when there are prim::profile nodes in
the graph. These two passes run next to each other, so we could simply
remove prim::profile nodes slightly earlier: before the BatchMM pass.

Differential Revision: [D23453266](https://our.internmc.facebook.com/intern/diff/D23453266)

[ghstack-poisoned]
